hermes/probe/monitor.go

```diff
@@ -12,9 +12,164 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
+// Authors:
+// - Alicja Kwiecinska (kwiecinskaa@google.com) github: alicjakwie
+// - Evan Spendlove, GitHub: evanSpendlove.
+//
 // Monitor implements a probe that monitors a storage system using the
 // Hermes algorithm.
+// Package probe implements the probe that Hermes uses to monitor
+// a storage system.
+
+package probe
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+	"github.com/googleinterns/step224-2020/hermes"
+
+	probepb "github.com/googleinterns/step224-2020/config/proto"
+	hermespb "github.com/googleinterns/step224-2020/hermes/proto"
+)
 
 // MonitorProbe holds aggregate information about all probe runs, per-target.
 // It also holds the config and options used to initialise the probe.
-type MonitorProbe struct{}
+type MonitorProbe struct {
```

Comment:
It seems like in here you are collecting a lot of information that is already available in other places (e.g. HermesProbeDef already contains targets information). At the moment it is a bit confusing whether it operates at the target level (like your comment suggests) or the hermes.Hermes level. I think it would make sense to

1. Move/collapse/combine `Config`, `opts`, and `targets` to/into the Hermes struct. 
2. Since the intention is to run on a per-target basis, you could instead make a `Target` struct elsewhere that takes a `*probe.Target`, `LatencyMetric`, and mutex (plus whatever you think its relevant to monitor). Hermes would have a `[]*Targets` whose `*probe.Target` would either match the ones from Config, or be the exact same  thing.
3. At this stage you can decide whether you want to move the monitoring responsibilities to hermes.go or to keep it separate and have it take a Hermes struct as a field like you do now (in this file).

Reply:
Thank you for raising this! I've thought about this and I think using a `Target` struct to encapsulate the information and state required for a target run is the best solution. It removes the need for a shared map of metrics per target and the duplication of the `mp.targets` field. It also allows us to store the state (via `StateJournal`) of the target bucket conveniently.

I think this also removes the need for a mutex lock as we no longer have fields that require concurrent access.